static/app/utils/withProjectsSpecified.tsx

```diff
@@ -24,27 +21,33 @@ type State = {
 /**
  * Higher order component that takes specificProjectSlugs and provides list of that projects from ProjectsStore
  */
-const withProjectsSpecified = <P extends InjectedProjectsProps>(
+function withProjectsSpecified<P extends InjectedProjectsProps>(
   WrappedComponent: React.ComponentType<P>
-) =>
-  createReactClass<Props & Omit<P, keyof InjectedProjectsProps>, State>({
-    displayName: `withProjectsSpecified(${getDisplayName(WrappedComponent)})`,
-    mixins: [Reflux.listenTo(ProjectsStore, 'onProjectUpdate') as any],
-    getInitialState() {
-      return ProjectsStore.getState(this.props.specificProjectSlugs);
-    },
-
-    UNSAFE_componentWillReceiveProps(nextProps: Props) {
-      const {specificProjectSlugs} = this.props;
-
-      if (xor(nextProps.specificProjectSlugs, specificProjectSlugs).length) {
-        this.setState(ProjectsStore.getState(nextProps.specificProjectSlugs));
+) {
+  class WithProjectsSpecified extends React.Component<
+    Props & Omit<P, keyof InjectedProjectsProps>,
+    State
+  > {
+    static displayName = `withProjectsSpecified(${getDisplayName(WrappedComponent)})`;
+
+    state = ProjectsStore.getState(this.props.specificProjectSlugs);
+
+    static getDerivedStateFromProps(nextProps: Readonly<Props>): State {
+      return ProjectsStore.getState(nextProps.specificProjectSlugs);
+    }
+
+    componentWillUnmount() {
+      this.unsubscribe();
+    }
+
+    unsubscribe = ProjectsStore.listen(() => {
+      const storeState = ProjectsStore.getState(this.props.specificProjectSlugs);
+
+      if (this.state !== storeState) {
```

Comment:
Do you need a deep equals here or is the shallow comparison going to be ok?

Reply:
I thought about it when refactoring but from my testing, the shallow one did not matter.
I'll add `isEqual` just to be super explicit, it should not affect performance as this does not get fired often.